proxy/inetaf_tcpproxy_integration_test.go

```diff
@@ -0,0 +1,177 @@
+package proxy
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"testing"
+
+	"github.com/fabiolb/fabio/config"
+	"github.com/fabiolb/fabio/proxy/tcp"
+	"github.com/fabiolb/fabio/proxy/tcp/tcptest"
+	"github.com/fabiolb/fabio/route"
+)
+
+// to run this test, add the following to /etc/hosts:
+// 127.0.0.1	example.com
+// 127.0.0.1	example2.com
+// and then set the environment FABIO_IHAVEHOSTENTRIES=true
+// This also runs in TRAVIS by default, since .travis.yml adds these aliases.
+func TestProxyTCPAndHTTPS(t *testing.T) {
+	if os.Getenv("TRAVIS") != "true" && os.Getenv("FABIO_IHAVEHOSTENTRIES") != "true" {
+		t.Skip("skipping because env FABIO_IHAVEHOSTENTRIES is not set to true")
+	}
+
+	tlsCfg1 := tlsServerConfig()
+	tlsCfg2 := tlsServerConfig2()
+	tcpServer := httptest.NewUnstartedServer(okHandler)
+	tcpServer.TLS = tlsCfg2
+	tcpServer.StartTLS()
+	defer tcpServer.Close()
+
+	httpPayload := []byte(`OK HTTP`)
+
+	httpServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Write(httpPayload)
+	}))
+	defer httpServer.Close()
+
+	tpl := `route add srv / %s opts "proto=https"
+route add tcproute example2.com/ tcp://%s opts "proto=tcp"`
+
+	table, _ := route.NewTable(bytes.NewBufferString(fmt.Sprintf(tpl, httpServer.URL, tcpServer.Listener.Addr())))
+	hp := &HTTPProxy{
+		Lookup: func(r *http.Request) *route.Target {
+			return table.Lookup(r, "", route.Picker["rr"], route.Matcher["prefix"], globCache, globEnabled)
+		},
+	}
+
+	tp := &tcp.SNIProxy{
+		Lookup: func(h string) *route.Target {
+			return table.LookupHost(h, route.Picker["rr"])
+		},
+	}
+	m := func(_ context.Context, h string) bool {
+		// TODO - matcher needs to move out of main
+		// so we can test it more easily.  Probably
+		// the other functions too.
+		t := table.LookupHost(h, route.Picker["rr"])
+		if t == nil {
+			return false
+		}
+		// Make sure this is supposed to be a tcp proxy.
+		// opts proto= overrides scheme if present.
+		var (
+			ok    bool
+			proto string
+		)
+		if proto, ok = t.Opts["proto"]; !ok && t.URL != nil {
+			proto = t.URL.Scheme
+		}
+		return "tcp" == proto
+	}
+
+	// get an unused port for use for the proxy.  the rest of the tests just
+	// pick a high-numbered port, but this should be safer, if ugly.  could
+	// also just fire up a listener with 0 as the port and let the stack
```

Comment:
this is what hashi used to do for consul testing, as well, however it doesn't work on all os's (this is actually one of the first things @magiconair fixed properly when he consulted for them)

old: https://github.com/hashicorp/consul/blob/v0.8.5/testutil/server.go#L132

current: https://github.com/hashicorp/consul/tree/v1.8.3/sdk/freeport

while i don't think its worth messing with now, may be nice to keep in the back of our minds if a user runs into problems while executing tests.

Reply:
your after link doesn't work, but I'd be interested to see what they come up with.  This test skips unless you explicitly create the /etc/hosts entries and set the environment variable mentioned in the comments above, but as far as I'm aware this should work cross platform provided it resolves example.com and example2.com to localhost.  So are you saying httptest is broken on some platforms?

Reply:
@nathanejohnson: it works for me...?

Reply:
I see the package, I assumed you were trying to point to a specific section of code as in the first link?  Also, is httptest broken on some platforms?  Because that's what I use to grab a free port

Reply:
No, I linked the package as it contains os-specific implementations. And no, i haven't encountered it personally.